variety/AddWallhavenDialog.py

```diff
@@ -0,0 +1,192 @@
+# with this program.  If not, see <http://www.gnu.org/licenses/>.
+### END LICENSE
+import urllib
+from urllib.parse import urlencode
+
+from gi.repository import Gdk, Gtk  # pylint: disable=E0611
+
+from variety.AddConfigurableDialog import AddConfigurableDialog
+from variety.plugins.builtin.downloaders.WallhavenDownloader import WallhavenDownloader
+from variety.Util import Util
+from variety_lib.helpers import get_builder
+
+
+class AddWallhavenDialog(AddConfigurableDialog):
+    __gtype_name__ = "AddWallhavenDialog"
+
+    def __new__(cls):
+        builder = get_builder("AddWallhavenDialog")
+        new_object = builder.get_object("add_wallhaven_dialog")
+        new_object.finish_initializing(builder)
+        return new_object
+
+    def finish_initializing(self, builder):
+        """Called when we're finished initializing.
+
+        finish_initalizing should be called after parsing the ui definition
+        and creating a WelcomeDialog object with it in order to
+        finish initializing the start of the new WelcomeDialog
+        instance.
+        """
+        # Get a reference to the builder and set up the signals.
+        self.builder = builder
+        self.ui = builder.get_ui(self)
+        self.edited_row = None
+        self.load_ui()
+
+    def set_source(self, source):
+        super().set_source(source)
+        self.ui.use_apikey_or_not.set_label(source.get_ui_use_apikey_or_not_text())
+        self.ui.getting_api_key_instruction.set_markup(source.get_api_key_instruction())
+        self.ui.apikey_instruction.set_text(source.get_ui_apikey_instruction())
+
+    def set_edited_row(self, edited_row):
+        super().set_edited_row(edited_row)
+        original_location = self.edited_row[2]
+
+        location = WallhavenDownloader(None, original_location).url
+        url_parts, params = self.parse_location(location)
+        if "usekey" in params:
+            self.ui.use_apikey_or_not.set_active(params["usekey"][0])
+        if "apikey" in params:
+            self.ui.apikey.set_text(params["apikey"][0])
+
+    @staticmethod
+    def parse_location(location):
+        url_parts = urllib.parse.urlparse(location)
+        params = urllib.parse.parse_qs(url_parts.query)
+
+        return url_parts, params
+
+    def ok_thread(self):
+        def _start_ui():
+            self.ui.buttonbox.set_sensitive(False)
+            self.ui.message.set_visible(True)
+            self.ui.query.set_sensitive(False)
+            self.ui.spinner.set_visible(True)
+            self.ui.spinner.start()
+            self.ui.error.set_label("")
+            self.ui.use_apikey_or_not.set_sensitive(False)
+            self.ui.apikey.set_sensitive(False)
+
+        Util.add_mainloop_task(_start_ui)
+
+        query = self.ui.query.get_text().strip()
+        query = WallhavenDownloader(None, query).web_url
+        usekey = self.ui.use_apikey_or_not.get_active()
+
+        valition_empty_apikey = False
```

Comment:
Typo: "validation_empty_apikey"? But if we remove the checkbox and just check if an api key is specified or not, we won't need the validation logic below anyway.

Reply:
>Typo: "validation_empty_apikey"?

Yes

>But if we remove the checkbox and just check if an api key is specified or not, we won't need the validation logic below anyway.

Good idea